### PR TITLE
poc kubeapps api service

### PR DESCRIPTION
**Pull request by absoludity**
### Description of the change

All the info for trying out the process of using/updating the API is in the [README.md](https://github.com/absoludity/kubeapps/blob/poc-kubeapps-api-service/cmd/kubeapps-api-service/README.md). The other options considered are documented in [Kubeapps: pluggable support for other package formats](https://docs.google.com/document/d/1FxWjrGMK1r3zpNlSASHQyH4SHl5dGf6gWgQpcRauPSw/edit#).

I'll use this as a base to propose implementation of the stubbed methods next week, so we can see what's involved in modifying or implementing APIs.

### CODE REVIEW

**Review by andresmgot — COMMENTED** [review]
Neat combination of grpc and the plugin system! Let's see how this evolves.

One question, at the moment, this seems to target packages only (like charts) but the goal is for this API service to define the API for instances as well (like releases), right?

File: cmd/kubeapps-api-service/README.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+## Plug-able
+
+After trying a number of methods and weighing those up with the pros and cons outlined in the design doc (I'll make that public next week), I've used the standard [go plugin package](https://golang.org/pkg/plugin/).
+
+Each plugin consists of 3 source files (and some generated files):
+
+* A `.proto` file defining the service that uses the messages defined in our core,
+* A `server.go` that implements the actual function(s), and
+* A `plugin/main.go` that is used to build the `.so` file for that plugin. This `main.go` has two public functions: one to register the plugin with a GRPC server and one to register the plugin for the http handler.
```
[the remark this answers is not in the document]
Comment (absoludity) [reply cmd/kubeapps-api-service/README.md:27]: > Neat combination of grpc and the plugin system! Let's see how this evolves.
> 
> One question, at the moment, this seems to target packages only (like charts) but the goal is for this API service to define the API for instances as well (like releases), right?

Yes, that's exactly what the two API stubs are that I've defined so far are meant to be:

* `PackageRepositories` is for adding and querying repositories, such as getting available packages, while
* `Packages` is for adding and querying the installed packages (releases)

I decided not to use the helm specific term `Releases` and instead follow the naming convention of `InstalledPackage` that VMware seems to be adopting across the board.

Thanks for the input so far!

[the remark this answers is not in the document]
Comment (andresmgot) [reply cmd/kubeapps-api-service/README.md:27]: > Yes, that's exactly what the two API stubs are that I've defined so far are meant to be:
> 
> * `PackageRepositories` is for adding and querying repositories, such as getting available packages, while
> * `Packages` is for adding and querying the installed packages (releases)
> 
> I decided not to use the helm specific term `Releases` and instead follow the naming convention of `InstalledPackage` that VMware seems to be adopting across the board.
> 

I am not sure I understood this. `PackageRepositories` are the current packages while `Packages` are instances? That can be quite confusing. 

I am not sure what naming convention you are referring but doesn't make more sense to have:
 - Repository: Catalog of packages
 - Package: App definition
 - Instance: Instantiated package 

?

File: cmd/kubeapps-api-service/README.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+This means that the actual go command just needs to load all the `.so` files from the specified plugin dirs and register them to start. You can see this in the `kubeapps-api-service/server/server.go` file.
+
+**Note**: I've chosen to demo both simple RPC calls (in the `kubeappsapis/core/v1/core.proto` you'll see the `PluginsAvailable` which returns a simple `PluginsAvailableResponse`) while the `GetAvailablePackages` call in `kubeappsapis/plugins/packagerepositories/helm/v1/helm.proto` returns a streaming response - no more need for pagination :) For more info about these two types of RPC calls, see [core-concepts](https://grpc.io/docs/what-is-grpc/core-concepts/).
```
[the remark this answers is not in the document]
Comment (andresmgot) [reply cmd/kubeapps-api-service/README.md:31]: :+1: thanks for the clarification. Yeah, I thought that pagination was still needed, basically to not request every logo (for example) every-time the catalog is requested.

File: cmd/kubeapps-api-service/cmd/root.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		// Find home directory.
+		home, err := homedir.Dir()
+		cobra.CheckErr(err)
+
+		// Search config in home directory with name ".kubeapps-api-service" (without extension).
+		viper.AddConfigPath(home)
+		viper.SetConfigName(".kubeapps-api-service")
+	}
+
+	viper.AutomaticEnv() // read in environment variables that match
```
[the remark this answers is not in the document]
Comment (andresmgot) [reply cmd/kubeapps-api-service/cmd/root.go:78]: yes, but I don't know what can be configured with this. I guess that whatever is in the config file can be set with env vars but what is configurable, the IP/port or something related to the plugins?

File: cmd/kubeapps-api-service/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (absoludity) [reply cmd/kubeapps-api-service/README.md:27]: > > I decided not to use the helm specific term `Releases` and instead follow the naming convention of `InstalledPackage` that VMware seems to be adopting across the board.
> 
> I am not sure I understood this. `PackageRepositories` are the current packages while `Packages` are instances? That can be quite confusing.

`PackageRepositories` are the repositories which define what `Packages` are available (I agree it's confusing, should be `AvailablePackages` or something indicating, but not my decision).

> 
> I am not sure what naming convention you are referring but doesn't make more sense to have:

I'll find the doc and ping you internally. I need to go over it again too.

File: cmd/kubeapps-api-service/cmd/root.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (absoludity) [reply cmd/kubeapps-api-service/cmd/root.go:78]: Yes, current config options are the port as well as the (multiple) `--plugin-dir`s which are scanned for plugins. See the actual server.go code I've highlighted below.

**Pull request closed without merging**